File: C++/Solution987.cpp
#include<iostream>
#include<vector>
#include<queue>

using namespace std;

typedef vector<int> VI;
typedef vector<VI> VII;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

struct cmp {
    bool operator()(vector<int>& vec1, vector<int>& vec2){
        for(int i = 0; i < 3; i++){
            if(vec1[i] != vec2[i]){
                return vec1[i] > vec2[i];
            }
        }
        return false;
    }
};

class Solution987 {
public:
    vector<vector<int>> verticalTraversal(TreeNode* root) {
        dfs(root, 0, 0);
        vector<vector<int>> ans;
        vector<int> vec;
        if(!root){
            return ans;
        }
        int col = pq.top()[0];
        while(!pq.empty()){
            auto ele = pq.top();
            pq.pop();
            if(ele[0] == col){
                vec.push_back(ele[2]);
            }
            else{
                ans.push_back(vec);
                vec.resize(0);
                vec.push_back(ele[2]);
                col = ele[0];
            }
        }
        ans.push_back(vec);
        return ans;
    }

private:
    priority_queue<VI, VII, cmp> pq;

    void dfs(TreeNode* root, int row, int col){
        if(!root){
            return;
        }
        pq.push({col, row, root -> val});
        dfs(root -> left, row + 1, col - 1);
        dfs(root -> right, row + 1, col + 1);
    }
};

File: C++/Solution987.java
import java.util.*;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
      this.val = val;
      this.left = left;
      this.right = right;
    }
}
 


class Solution987 {
    private PriorityQueue<int[]> pq;

    public List<List<Integer>> verticalTraversal(TreeNode root) {
        pq = new PriorityQueue<>(new Comparator<>(){
            @Override
            public int compare(int[] x, int[] y){
                for(int i = 0; i < 3; i++){
                    if(x[i] != y[i]){
                        return x[i] - y[i];
                    }
                }
                return -1;
            }
        }
        );
        dfs(root, 0, 0);
        List<List<Integer>> ans = new LinkedList<>();
        List<Integer> list = new LinkedList<>();
        if(root == null){
            return ans;
        }
        int col = pq.peek()[0];
        while(!pq.isEmpty()){
            int[] cur = pq.poll();
            if(cur[0] == col){
                list.add(cur[2]);
            }
            else{
                ans.add(list);
                list = new LinkedList<>();
                list.add(cur[2]);
                col = cur[0];
            }
        }
        ans.add(list);
        return ans;
    }

    private void dfs(TreeNode root, int row, int col){
        if(root == null){
            return;
        }
        pq.offer(new int[]{col, row, root.val});
        dfs(root.left, row + 1, col - 1);
        dfs(root.right, row + 1, col + 1);
    }
}